### PR TITLE
Draft documentation on how to test extensions.

### PR DESCRIPTION
**Pull request by jcass77**
Contains the proposed documentation improvements outlined in #503.

### CODE REVIEW

**Comment by jcass77** [comment]
I pushed a small commit that fixes the capitalisation of some headings. Build failure seems to be as a result of an old flake8 violation (which I thought I saw @jodal fix on the master branch a few days ago?).

**Comment by jodal** [comment]
master and release-1.1 isn't merged into develop yet, ref. #1400. I cherry-picked the flake8-fixing commit now and will rerun the tests here.

**Comment by jcass77** [comment]
I just saw that the bullet lists are not rendering correctly after making a last-minute change last night. Let me fix that quick...

**Comment by jcass77** [comment]
Ok, I think the PR is ready for review now.

File: docs/extensiondev.rst [hunk 1]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+take a look at the test cases that ship with Mopidy itself which covers
+everything from instantiating various controllers, reading configuration files,
+and simulating events that your extension can listen to.
+
+In general your tests should cover the extension definition, the relevant
+Mopidy controllers, and the Pykka backend and / or frontend actors that form
+part of the extension.
+
+Testing the extension definition
+--------------------------------
```
Comment (jodal) [inline docs/extensiondev.rst]: I believe a blank line is needed after the header.

File: docs/extensiondev.rst [hunk 2]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+Test cases for checking the definition of the extension should ensure that:
```
Comment (jodal) [inline docs/extensiondev.rst]: I believe a blank line is needed before and after lists.

File: docs/extensiondev.rst [hunk 3]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+- that the config schema is fully defined, and
+- that the extension's actor(s) are added to the Mopidy registry on setup.
+
+An example of what these tests could look like is provided below::
+
+    def test_get_default_config(self):
+        ext = Extension()
+        config = ext.get_default_config()
+
+        self.assertIn('[my_extension]', config)
```
Comment (jodal) [inline docs/extensiondev.rst]: We should use pytest's assert syntax instead, e.g. `assert '[my_extension]' in config`.

File: docs/extensiondev.rst [hunk 4]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+
+        ext = Extension()
+        ext.setup(registry)
+        calls = [mock.call('frontend', frontend_lib.MyFrontend),
+                 mock.call('backend',  backend_lib.MyBackend)]
+        registry.add.assert_has_calls(calls, any_order=True)
+
+
+Testing backend actors
+----------------------
```
Comment (jodal) [inline docs/extensiondev.rst]: A newline here as well.

File: docs/extensiondev.rst [hunk 5]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+Backends can usually be constructed with a small mockup of the configuration
+file, and mocking the audio actor::
+
+    @pytest.fixture()
```
Comment (jodal) [inline docs/extensiondev.rst]: Should there be `()` after `fixture`?

File: docs/extensiondev.rst [hunk 6]
```diff
@@ -542,3 +542,227 @@ your HTTP requests::
...
+You'll probably want to patch ``requests`` or any other web API's that you use
+to avoid any unintended HTTP requests from being made by your backend during
+testing::
+
+    from mock import patch
+    @mock.patch('requests.get',
+                mock.Mock(side_effect=Exception('Intercepted unintended HTTP call')))
+
+
+Backend tests should also ensure that:
```
Comment (jodal) [inline docs/extensiondev.rst]: Newline before list.

File: docs/extensiondev.rst [hunk 7]
```diff
@@ -542,3 +542,229 @@ your HTTP requests::
...
+
+
+You'll probably want to patch ``requests`` or any other web API's that you use
+to avoid any unintended HTTP requests from being made by your backend during
+testing::
+
+    from mock import patch
+    @mock.patch('requests.get',
+                mock.Mock(side_effect=Exception('Intercepted unintended HTTP call')))
+
```
Comment (jodal) [inline docs/extensiondev.rst]: Should we mention [responses](https://pypi.python.org/pypi/responses) or [vcrpy](https://pypi.python.org/pypi/vcrpy) here?

File: docs/extensiondev.rst (hunk 5 above)
Comment (jcass77) [inline docs/extensiondev.rst]: Both probably work but I think this was left over after initially being ```@pytest.fixture(scope='session')```.

File: docs/extensiondev.rst [hunk 8]
```diff
@@ -542,3 +542,229 @@ your HTTP requests::
...
+                config, backends=[get_backend(config]).proxy()
+
+
+It may be advisable to take a quick look at the
+`Pykka API <https://www.pykka.org/en/latest/>`_ at this point to make sure that
+you are familiar with ``ThreadingActor``, ``ThreadingFuture``, and the
+``proxies`` that allow you to access the attributes and methods of the actor
+directly.
+
+You'll also need a list of ``models.Track`` and a list of URIs in order to
```
Comment (jodal) [inline docs/extensiondev.rst]: Use this instead to link to the Track API docs:

```
:class:`~mopidy.models.Track`
```

**Comment by jcass77** [comment]
Thank you for the pointers. All of the review comments have been addressed in https://github.com/jcass77/mopidy/commit/edc3929dafe73dfb4f7d2b59711c7ab57b2df618.

**Comment by jodal** [comment]
Thanks :-)

File: docs/extensiondev.rst [hunk 9]
```diff
@@ -542,3 +542,245 @@ your HTTP requests::
...
+audio URL to Mopidy's core for playback.
+
+Testing frontends
+-----------------
+
+Because most frontends will interact with the Mopidy core, it will most likely
+be necessary to have a full core running for testing purposes::
+
+    self.core = core.Core.start(
+                config, backends=[get_backend(config]).proxy()
```
Comment (ZenithDK) [inline docs/extensiondev.rst]: There's a mix-up here in the parenthesis and the square bracket.
It should be "[get_backend(config).proxy()]"

Comment (jcass77) [inline docs/extensiondev.rst]: Yes there is a mix-up. It should actually be:
```
    self.core = core.Core.start(
                config, backends=[get_backend(config)].proxy()

Thanks

Comment (jcass77) [inline docs/extensiondev.rst]: Actually,
```
    self.core = core.Core.start(
                config, backends=[get_backend(config)]).proxy()
```
What a tongue twister.

Comment (jodal) [inline docs/extensiondev.rst]: Both of you have commit access. Please fix :-)

Comment (jcass77) [inline docs/extensiondev.rst]: What is the process for collaborators? Should we still create PR's and then just merge them ourselves or how does the peer review procedure work in such instances?

Comment (jodal) [inline docs/extensiondev.rst]: For simple doc fixes, just commit.

**Pull request merged**